Try PULL image more then one if service response HTTP 505

If the docker registry returns HTTP 503, I would prefer to run docker pull once more.
Sometimes, a second run works properly.

What do you think? @pkubatrh @hhorak @praiskup.

See the error:
```
make[1]: Entering directory `/tmp/daily_scl_tests/s2i-nodejs-container'
mkdir -p 10/root
go-md2man -in "10/README.md" -out "10/root/help.1"
chmod a+r "10/root/help.1"
VERSION="10" SKIP_SQUASH=1 UPDATE_BASE= OS=fedora CLEAN_AFTER= DOCKER_BUILD_CONTEXT=. OPENSHIFT_NAMESPACES="" CUSTOM_REPO="" /usr/bin/env bash common/build.sh
-> Version 10: building image from 'Dockerfile.fedora' ...
Sending build context to Docker daemon 50.69 kB

Step 1/13 : FROM registry.fedoraproject.org/f29/s2i-base:latest
Trying to pull repository registry.fedoraproject.org/f29/s2i-base ... 
latest: Pulling from registry.fedoraproject.org/f29/s2i-base
7692efc5f81c: Pulling fs layer
aaf5ad2e1aa3: Pulling fs layer
05ba4d604a56: Pulling fs layer
error pulling image configuration: received unexpected HTTP status: 503 Service Temporarily Unavailable
make[1]: *** [10] Error 1
make[1]: Leaving directory `/tmp/daily_scl_tests/s2i-nodejs-container'
make: *** [build-serial] Error 2
```

--- a/build.sh
+++ b/build.sh
@@ -86,21 +86,17 @@ function pull_image {
       continue
     fi
 
-    image_name=$(echo "$line" | cut -d ' ' -f2)
     # In case FROM scratch is defined, skip it
     if grep -q "scratch" <<< "$image_name"; then
-        continue
+      continue
     fi
-    # In case FROM ubi8 as ubi-base is mentioned
-    # pull ubi8 image
-    if grep -q " as " <<< "$line"; then
-        image_name=$(echo "$line" | cut -d ' ' -f2)
-    fi
+    image_name=$(echo "$line" | cut -d ' ' -f2)
     echo "-> Pulling image $image_name before building image from $dockerfile."
     # Sometimes in Fedora case it fails with HTTP 50X
     # Check if the image is available locally and try to pull it if it is not
-    if docker images "$image_name" &>/dev/null; then
-        continue
+    if [[ "$(docker images -q "$image_name" 2>/dev/null)" != "" ]]; then
+      echo "The image $image_name is already pulled."
+      continue
     fi
     # Try pulling the image to see if it is accessible
     if ! docker pull "$image_name"; then

--- a/build.sh
+++ b/build.sh
@@ -86,11 +86,12 @@ function pull_image {
       continue
     fi
 
+    image_name=$(echo "$line" | cut -d ' ' -f2)
+
     # In case FROM scratch is defined, skip it
-    if grep -q "scratch" <<< "$image_name"; then
+    if [[ x"$image_name" == "xscratch" ]]; then
       continue
     fi
-    image_name=$(echo "$line" | cut -d ' ' -f2)
     echo "-> Pulling image $image_name before building image from $dockerfile."
     # Sometimes in Fedora case it fails with HTTP 50X
     # Check if the image is available locally and try to pull it if it is not

--- a/build.sh
+++ b/build.sh
@@ -76,6 +76,37 @@ function clean_image {
   fi
 }
 
+# Pull image based on FROM, before we build our own.
+function pull_image {
+  local dockerfile="$1"
+
+  # Get image_name from Dockerfile before pulling.
+  while read -r line; do
+    if ! grep -q "^FROM" <<< "$line"; then
+      continue
+    fi
+
+    image_name=$(echo "$line" | cut -d ' ' -f2)
+    # In case FROM scratch is defined, skip it
+    if grep -q "scratch" <<< "$image_name"; then
+        continue
+    fi
+    # In case FROM ubi8 as ubi-base is mentioned
+    # pull ubi8 image
+    if grep -q " as " <<< "$line"; then
+        image_name=$(echo "$line" | cut -d ' ' -f2)
+    fi
+    echo "-> Pulling image $image_name before building image from $dockerfile."
+    # Try pulling the image to see if it is accessible
+    # Sometimes in Fedora case it fails with HTTP 50X
+    if ! docker pull "$image_name"; then
+      echo "Pulling image $image_name failed. Let's wait 2 seconds and try one more time."
+      sleep 2
+      docker pull "$image_name"
+    fi
+  done < "$dockerfile"
+}
+
 # Perform docker build but append the LABEL with GIT commit id at the end
 function docker_build_with_version {
   local dockerfile="$1"
@@ -106,6 +137,8 @@ function docker_build_with_version {
       echo "ERROR: file type not known: $CUSTOM_REPO" >&2
     fi
   fi
+
+  pull_image "$dockerfile"
 
   # shellcheck disable=SC2016
   parse_output 'docker build '"$BUILD_OPTIONS"' -f "$dockerfile" "${DOCKER_BUILD_CONTEXT}"' \

--- a/build.sh
+++ b/build.sh
@@ -97,13 +97,18 @@ function pull_image {
         image_name=$(echo "$line" | cut -d ' ' -f2)
     fi
     echo "-> Pulling image $image_name before building image from $dockerfile."
-    # Try pulling the image to see if it is accessible
     # Sometimes in Fedora case it fails with HTTP 50X
+    # Check if the image is available locally and try to pull it if it is not
+    if docker images "$image_name" &>/dev/null; then
+        continue
+    fi
+    # Try pulling the image to see if it is accessible
     if ! docker pull "$image_name"; then
       echo "Pulling image $image_name failed. Let's wait 2 seconds and try one more time."
       sleep 2
       docker pull "$image_name"
     fi
+
   done < "$dockerfile"
 }
 

